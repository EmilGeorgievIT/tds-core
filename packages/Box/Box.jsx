import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import safeRest from '../../shared/utils/safeRest'
import joinClassNames from '../../shared/utils/joinClassNames'
import capitalize from '../../shared/utils/capitalize'

const BoxTag = ({
  tag,
  vertical,
  horizontal,
  inset,
  below,
  between,
  inline,
  className,
  children,
}) => {
  return React.createElement(tag, { className: className }, children)
}

const spacingBase = 1 // 16px

const mobileSize = {
  1: spacingBase * 0.25 + 'rem',
  2: spacingBase * 0.5 + 'rem',
  3: spacingBase + 'rem',
  4: spacingBase * 1.5 + 'rem',
  5: spacingBase * 2 + 'rem',
  6: spacingBase * 2.5 + 'rem',
  7: spacingBase * 3 + 'rem',
  8: spacingBase * 4 + 'rem',
}

const desktopSize = {
  1: spacingBase * 0.25 + 'rem',
  2: spacingBase * 0.5 + 'rem',
  3: spacingBase + 'rem',
  4: spacingBase * 2 + 'rem',
  5: spacingBase * 3 + 'rem',
  6: spacingBase * 4 + 'rem',
  7: spacingBase * 4.5 + 'rem',
  8: spacingBase * 6 + 'rem',
}

const breakpoints = {
  sm: 576,
  md: 768,
  lg: 992,
  xl: 1200,
}

const StyledBoxTag = styled(BoxTag)`
  display: ${props => (props.between ? 'flex' : 'initial')};
  flex-direction: ${props =>
    props.inline && props.between ? 'row' : props.between ? 'column' : 'initial'};
  padding-top: ${props => (props.inset ? mobileSize[props.inset] : mobileSize[props.vertical])};
  padding-bottom: ${props => (props.inset ? mobileSize[props.inset] : mobileSize[props.vertical])};
  padding-left: ${props => (props.inset ? mobileSize[props.inset] : mobileSize[props.horizontal])};
  padding-right: ${props => (props.inset ? mobileSize[props.inset] : mobileSize[props.horizontal])};
  margin-bottom: ${props =>
    !props.inline && props.below && !props.between ? mobileSize[props.below] : '0rem'};

  > *:not(:last-child) {
    margin-bottom: ${props =>
      !props.inline && !props.below && props.between ? mobileSize[props.between] : '0rem'};
    margin-right: ${props => (props.inline && props.between ? mobileSize[props.between] : '0rem')};
  }

  @media only screen and (min-width: ${breakpoints.md}px) {
    padding-top: ${props => (props.inset ? desktopSize[props.inset] : desktopSize[props.vertical])};
    padding-bottom: ${props =>
      props.inset ? desktopSize[props.inset] : desktopSize[props.vertical]};
    padding-left: ${props =>
      props.inset ? desktopSize[props.inset] : desktopSize[props.horizontal]};
    padding-right: ${props =>
      props.inset ? desktopSize[props.inset] : desktopSize[props.horizontal]};
    margin-bottom: ${props =>
      !props.inline && props.below && !props.between ? desktopSize[props.below] : '0rem'};

    > *:not(:last-child) {
      margin-bottom: ${props =>
        !props.inline && !props.below && props.between ? desktopSize[props.between] : '0rem'};
      margin-right: ${props =>
        props.inline && props.between ? desktopSize[props.between] : '0rem'};
    }
  }
`

/**
 * Apply spacing within or around components.
 *
 * @version ./package.json
 */
const Box = ({
  tag,
  vertical,
  horizontal,
  inset,
  below,
  between,
  inline,
  dangerouslyAddClassName,
  children,
  ...rest
}) => {
  return (
    <StyledBoxTag
      tag={tag}
      vertical={vertical}
      horizontal={horizontal}
      inset={inset}
      below={below}
      between={between}
      inline={inline}
    >
      {children}
    </StyledBoxTag>
  )
}

Box.propTypes = {
  /**
   * Specify an HTML element to render, such as `section`.
   */
  tag: PropTypes.string,
  /**
   * Indent content from the container's top and bottom edge by applying padding.
   */
  vertical: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8]),
  /**
   * Indent content from the container's left and right edge by applying padding.
   */
  horizontal: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8]),
  /**
   * Indent content from all of the container's edges by applying padding.
   */
  inset: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8]),
  /**
   * @ignore
   *
   * We are keeping this hidden for now as we are not sold on the necessity. We use it internally still to apply
   * spacing to Markdown components, but would like to use between instead if the library allows it.
   *
   * Sets a `margin-bottom`.
   */
  below: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8]),
  /**
   * Create either a block or an inline stack, applying margin in between every direct child. Margin will not be
   * applied to the last component in the stack.
   *
   * By default, `between` will arrange the Box's children as a flex column. Combine with `inline` to arrange them
   * as a flex row.
   */
  between: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8]),
  /**
   * Arrange children in a row. Combine with `between` to apply margins in between the row's elements.
   */
  inline: PropTypes.bool,
  /**
   * Append custom classes to `className`. Use sparingly, and do not attempt to override Box style properties as that
   * may cause unexpected behaviour.
   *
   * You would typically use this feature to apply flex alignment properties in combination with `between`.
   */
  dangerouslyAddClassName: PropTypes.string,
  /**
   * The content. Can be text, any HTML element, or any component.
   */
  children: PropTypes.node.isRequired,
}

Box.defaultProps = {
  inline: false,
  tag: 'div',
  vertical: undefined,
  horizontal: undefined,
  inset: undefined,
  below: undefined,
  between: undefined,
  dangerouslyAddClassName: undefined,
}

export default Box
