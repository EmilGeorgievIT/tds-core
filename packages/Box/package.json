{
  "name": "@tds/core-box",
  "version": "1.0.0",
  "main": "dist/index.cjs.js",
  "module": "dist/index.es.js",
  "license": "MIT",
  "peerDependencies": {
    "react": ">=15",
    "react-dom": ">=15"
  },
  "dependencies": {
    "@tds/core-responsive": "^1.0.0",
    "prop-types": "^15.5.10"
  }
}
