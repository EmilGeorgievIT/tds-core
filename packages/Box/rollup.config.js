import configure from '../../config/rollup.config'

export default configure({
  input: './Box.jsx',
  external: '@tds/core-responsive',
})
