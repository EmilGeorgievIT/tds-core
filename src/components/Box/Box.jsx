import React from 'react'
import PropTypes from 'prop-types'

import safeRest from '../../utils/safeRest'
import joinClassNames from '../../utils/joinClassNames'
import capitalize from '../../utils/capitalize'

import styles from './Box.modules.scss'

const getClassName = (spacing, location, scale) => {
  if (!scale) {
    return undefined
  }

  return styles[`${location}${capitalize(spacing)}-${scale}`]
}

const getBetweenClasses = (betweenSize, inline) => {
  if (!betweenSize) {
    return undefined
  } else if (inline) {
    return styles[`betweenRightMargin-${betweenSize}`]
  }

  return styles[`betweenBottomMargin-${betweenSize}`]
}

/**
 * Box - apply spacing within or around components
 */
const Box = ({
  inline,
  tag,
  vertical,
  horizontal,
  inset,
  below,
  right,
  left,
  between,
  dangerouslyAddClassName,
  children,
  ...rest
}) => {
  const xSize = inset || horizontal
  const ySize = inset || vertical

  let Tag = ''
  if (tag) {
    Tag = tag
  } else if (inline) {
    Tag = 'span'
  } else {
    Tag = 'div'
  }

  const classes = joinClassNames(
    getClassName('padding', 'horizontal', xSize),
    getClassName('padding', 'vertical', ySize),
    getClassName('margin', 'bottom', below),
    getClassName('margin', 'right', right),
    getClassName('margin', 'left', left),
    getBetweenClasses(between, inline),
    dangerouslyAddClassName
  )

  return React.createElement(
    Tag,
    {
      ...safeRest(rest),
      className: classes || undefined,
    },
    children
  )
}

Box.propTypes = {
  /**
   * If `true`, render a `<span>` instead of `<div>`.
   */
  inline: PropTypes.bool,
  /**
   *
   */
  tag: PropTypes.node,
  /**
   * Sets a `padding-top` and `padding-bottom`.
   */
  vertical: PropTypes.oneOf([1, 2, 3, 4, 6]),
  /**
   * Sets a `padding-left` and `padding-right`.
   */
  horizontal: PropTypes.oneOf([1, 2, 3, 4, 6]),
  /**
   * Sets `padding` on all four sides.
   */
  inset: PropTypes.oneOf([1, 2, 3, 4, 6]),
  /**
   * Sets a `margin-bottom`.
   */
  below: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8]),
  /**
   * Sets a `margin-right`.
   */
  right: PropTypes.oneOf([1, 2, 3, 4, 6]),
  /**
   * Sets a `margin-left`.
   */
  left: PropTypes.oneOf([1, 2, 3, 4, 6]),
  /**
   * Sets a `margin-bottom` to all direct children except the last. If the `inline` prop is set to `true`, this sets a `margin-right` to all direct children except the last.
   */
  between: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8]),
  /**
   * Append custom classes to `className`.
   */
  dangerouslyAddClassName: PropTypes.string,
  /**
   * The content. Can be text, any HTML element, or any component.
   */
  children: PropTypes.node.isRequired,
}

Box.defaultProps = {
  inline: false,
  tag: undefined,
  vertical: undefined,
  horizontal: undefined,
  inset: undefined,
  below: undefined,
  right: undefined,
  left: undefined,
  between: undefined,
  dangerouslyAddClassName: undefined,
}

export default Box
