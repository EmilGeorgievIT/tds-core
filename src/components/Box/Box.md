```jsx
<Card>
  <Box between={3}>
    <Heading level="h1">Our latest and greatest</Heading>
    <Paragraph>Everything you would expect, without conditions.</Paragraph>
    <ButtonLink href="//telus.com">Learn more</ButtonLink>
  </Box>
</Card>
```

## Spacing matrix

The Box component contains props that accept values from `1` to `8`. These numbered values correspond to our spacing system. Numbers `4` and above feature responsive spacing, as detailed in this table:

| Number | Default value | Value in viewports greater or equal to `768px` | 
| ------ | ------------: | -------------------------------------------: |
| 1 | 0.25rem (4px) | _unchanged_ |
| 2 | 0.5rem (8px) | _unchanged_ |
| 3 | 1rem (16px) | _unchanged_ |
| 4 | 1.5rem (24px) | 2rem (32px) |
| 5 | 2rem (32px) | 3rem (48px) |
| 6 | 2.5rem (40px) | 4rem (64px) |
| 7 | 3rem (48px) | 4.5rem (72px) |
| 8 | 4rem (64px) | 6rem (96px) |

### Usage tips

- Numbers `1` to `3` are suitable for spacing inline content such as [Typography](#typography), [Button](#button), and [ButtonLink](#buttonlink) components
- Numbers `4` to `8` are suitable for spacing major types of content such as [Cards](#card) and [Dividers](#dividers)
- Numbers `5`, `7`, and `8` are exclusively used to separating blocks vertically

### Example separating blocks

```jsx
<Box between={5}>
  <Box tag="section" inset={4} between={3}>
    <Heading level="h2">Lots to discover</Heading>
    <Paragraph>See our collection of home electronics.</Paragraph>
  </Box>

  <DimpleDivider/>

  <Box tag="section" inset={4} between={3}>
    <Heading level="h2">Something for everyone</Heading>
    <Paragraph>For you, your family and even your dog.</Paragraph>
  </Box>
</Box>
```

## Spacing in between elements

When wrapping a Box element around a series of elements, use the `between` prop to add either `margin-right` for inline elements or `margin-bottom` for stacked elements. Spacing will be applied to all directly-descending children except the last in the series.

### Stacked

```jsx 
<Box between={3}>
  <Box between={2}>
    <Heading level="h4">Wi-Fi Enabled</Heading>

    <Paragraph>Control any device connected to your SmartPlug from any location with internet access.</Paragraph>
  </Box>

  <Box between={2}>
    <Heading level="h4">Create ‘Scenes’</Heading>

    <Paragraph>Create your own scenes and rooms such as 'Kitchen' and manage all SmartPlugs in that scene.</Paragraph>
  </Box>

  <Box between={2}>
    <Heading level="h4">Compatibility</Heading>

    <Paragraph>Works with Google Assistant, Amazon Alexa, Apple HomeKit, and Nest.</Paragraph>
  </Box>
</Box>
```

### Inline

```jsx
<Box inline between={3}>
  <Link href="//telus.com">Phones</Link>
  <span>|</span>
  <Link href="//telus.com">Pre-Owned</Link>
  <span>|</span>
  <Link href="//telus.com">Prepaid</Link>
  <span>|</span>
  <Link href="//telus.com">Sale</Link>
</Box>
```
