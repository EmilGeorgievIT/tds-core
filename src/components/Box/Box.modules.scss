@import '../../scss/settings/responsive-grid';
@import '../../scss/utility/responsive';

$spacing-base: 1rem; // 16px

$spacings: (
  1: $spacing-base * 0.25,
  2: $spacing-base * 0.5,
  3: $spacing-base,
  4: $spacing-base * 1.5,
  5: $spacing-base * 2,
  6: $spacing-base * 2.5,
  7: $spacing-base * 3,
  8: $spacing-base * 4
);

$responsive-spacings: (
  medium:
    (
      4: $spacing-base * 2,
      5: $spacing-base * 3,
      6: $spacing-base * 4,
      7: $spacing-base * 4.5,
      8: $spacing-base * 6
    )
);

$boxProps: ('margin': 'Margin', 'padding': 'Padding');

// Todo remove unneeded classes related to margin
@mixin spacingClasses ($level, $value) {
  @each $p, $v in $boxProps {
    .vertical#{$v}-#{$level} {
      #{$p}-top: $value;
      #{$p}-bottom: $value;
    }

    .horizontal#{$v}-#{$level} {
      #{$p}-left: $value;
      #{$p}-right: $value;
    }

    .right#{$v}-#{$level} {
      #{$p}-right: $value;
    }

    .bottom#{$v}-#{$level} {
      #{$p}-bottom: $value;
    }

    // TODO: refactor to not generate unnecessary .betweenBottomPadding-*
    .betweenBottom#{$v}-#{$level} > *:not(:last-child) {
      #{$p}-bottom: $value;
    }

    .betweenRight#{$v}-#{$level} > *:not(:last-child) {
      #{$p}-right: $value;
    }

    .left#{$v}-#{$level} {
      #{$p}-left: $value;
    }
  }
}

@mixin responsiveSpacingClasses ($bp, $map) {
  @include from-breakpoint($bp) {
    @each $level, $value in $map {
      @include spacingClasses($level, $value);
    }
  }
}

@each $level, $value in $spacings {
  @include spacingClasses($level, $value);
}

@each $bp, $map in $responsive-spacings {
  @include responsiveSpacingClasses($bp, $map);
}
