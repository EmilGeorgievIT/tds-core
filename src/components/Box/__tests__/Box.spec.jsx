import React from 'react'
import { shallow } from 'enzyme'

import FlexBox from '../../Flexbox/Flexbox'
import Box from '../Box'

describe('Box', () => {
  const defaultProps = { between: 3 }
  const doShallow = (props = {}) =>
    shallow(
      <Box {...defaultProps} {...props}>
        Some content
      </Box>
    )

  it('renders', () => {
    const box = doShallow()

    expect(box).toMatchSnapshot()
  })

  it('can be either inline or block', () => {
    let box = doShallow()
    expect(box).toHaveTagName('div')

    box = doShallow({ inline: true })
    expect(box).toHaveClassName('betweenRight betweenRightMargin-3')
  })

  it('can render as other elements', () => {
    const box = doShallow({ tag: 'ul' })
    expect(box).toHaveTagName('ul')
  })

  it('can have vertical or horizontal spacing', () => {
    let box = doShallow({ vertical: 1 })
    expect(box).toHaveClassName('verticalPadding-1')

    box = doShallow({ horizontal: 1 })
    expect(box).toHaveClassName('horizontalPadding-1')
  })

  it('can have horizontal and vertical spacing of different scales', () => {
    const box = doShallow({ vertical: 1, horizontal: 2 })

    expect(box).toHaveClassName('verticalPadding-1 horizontalPadding-2')
  })

  it('can have inset padding', () => {
    const box = doShallow({ inset: 3 })

    expect(box).toHaveClassName('verticalPadding-3 horizontalPadding-3')
  })

  it('can have bottom margin', () => {
    const box = doShallow({ below: 2 })

    expect(box).toHaveClassName('bottomMargin-2')
  })

  it('can have right margin', () => {
    const box = doShallow({ right: 4 })

    expect(box).toHaveClassName('rightMargin-4')
  })

  it('can have between margin as a stack', () => {
    const box = doShallow({ between: 2 })

    expect(box).toHaveClassName('betweenBottomMargin-2')
  })

  it('can have between margin as inline', () => {
    const box = doShallow({ between: 2, inline: true })

    expect(box).toHaveClassName('betweenRight betweenRightMargin-2')
  })

  it('passes additional attributes to the element', () => {
    const box = doShallow({ id: 'the-id', 'data-some-attr': 'some value' })

    expect(box).toHaveProp('id', 'the-id')
    expect(box).toHaveProp('data-some-attr', 'some value')
  })

  it('will add additional arbitrary class names', () => {
    const box = doShallow({ dangerouslyAddClassName: 'a-class' })

    expect(box).toHaveClassName('a-class')
  })

  it('does not allow custom CSS', () => {
    const box = doShallow({ className: 'my-custom-class', style: { color: 'hotpink' } })

    expect(box).not.toHaveProp('className', 'my-custom-class')
    expect(box).not.toHaveProp('style')
  })
})
