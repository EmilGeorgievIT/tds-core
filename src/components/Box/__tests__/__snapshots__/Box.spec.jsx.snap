// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box renders 1`] = `
<div
  className="verticalPadding-2"
>
  Some content
</div>
`;
